### DISCUSSION TITLE
App Router setup without i18n routing

### DISCUSSION THREAD

**Discussion by lynn1286** [amannn/next-intl#1081] [Q&A]
App Router setup without i18n routing ，How to do get the language set by the user ？

```
import {NextIntlClientProvider} from 'next-intl';
import {getRequestConfig} from 'next-intl/server';
 
export default getRequestConfig(async () => {
  // This can either be defined statically if only a single locale
  // is supported, or alternatively read from the user settings
  const locale = 'en'; // -> How to get the language set by the user
 
  return {
    locale,
    messages: (await import(`../messages/${locale}.json`)).default
  };
});
```

**Comment by amannn** [comment]
It's entirely up to you in this case to store and retrieve the locale. Some options would be a user setting in your database, a custom cookie, a request header, etc. Is there anything that would work for your app?

**Reply by askjervold** [reply]
I actually came here yesterday looking for this exact discussion! We are also trying to implement next-intl without i18n routing, with user locale (and timezone) set through cookies using the [client hints pattern](https://github.com/epicweb-dev/epic-stack/blob/e20e5e1b18a62d793a4ead0a542dca65cb23fb9a/docs/client-hints.md). However, it wasn't clear to us what is actually available to us within the `getRequestConfig` invocation. To grab the cookies I would need access to the request, but I assume that's not available directly in `getRequestConfig` and that I would need to set it via middleware and pass it in here through params, but attempting to combine middleware and a setup without i18n routing seems to not be supported [according to the docs](https://next-intl-docs.vercel.app/docs/routing/middleware#troubleshooting)?

**Reply by amannn** [reply]
You can absolutely read from [`headers()`](https://nextjs.org/docs/app/api-reference/functions/headers) in `i18n.ts`! The function is executed as soon as a Server Component uses APIs from `next-intl` for the first time. So it's essentially the same as if you're calling `headers()` in a component. I think I should clarify this in the docs (edit: https://github.com/amannn/next-intl/commit/84de1350fb694b77cc2e60fb462057b430e3c4b8).

Here's an example that tries to use a cookie value, but falls back to using the `accept-language` header or a general fallback:


```tsx
import {getRequestConfig} from 'next-intl/server';
import {headers, cookies} from 'next/headers';
import {match} from '@formatjs/intl-localematcher';
import Negotiator from 'negotiator';

function resolveLocale() {
  const locales = ['en', 'de'];
  const defaultLocale = 'en';

  // Prio 1: Use value from a cookie
  let locale = cookies().get('your-custom-locale-cookie');
  
  // Prio 2: Use value from the `Accept-Language` header
  if (!locale) {
    const languages = new Negotiator({
      headers: {
        'accept-language': headers().get('accept-language') || undefined,
      },
    }).languages();
    try {
      locale = match(languages, locales, defaultLocale);
    } catch (e) {
      // Invalid language
    }
  }

  // Prio 3: Use the default locale
  if (!locale) {
    locale = defaultLocale;
  }

  return locale;
}

export default getRequestConfig(async () => {
  const locale = resolveLocale();
 
  return {
    locale,
    messages: (await import(`../messages/${locale}.json`)).default
  };
});
```

The time zone is indeed a bit tricker. Some options here are:
1. Use the server time zone (default)
2. Use the IP of the user (might be good for a default, but you should really let the user pick one explicitly if you do this)
3. Let the user pick one explicitly (persist to cookie or db)

Does that help?

It might be good to collect a few use cases like this, and potentially provide docs at some point (similar to the client hints pattern you've linked to)!

**Reply by askjervold** [reply]
That solves my problem perfectly! I'm still new to Next, so I wasn't aware I could access properties of the request here. I guess my next (heh) problem now is whether it is possible to change the locale without requiring a page refresh? I realize that would mean switching to client-side rendering, which I guess might not be possible to apply to the whole component tree without turning them into client components?

**Reply by amannn** [reply]
Nice! 👏 

> whether it is possible to change the locale without requiring a page refresh?

How are you storing your locale? In a cookie, like in the example above? You could try to invoke a Server Action that updates your cookie and then calls `revalidatePath('/')`. Theoretically, that should return updated markup for the page without a reload.

(I hope Next.js properly updates your cookie before the invalidation triggers the next render, otherwise you could try changing the cookie on the client side already and then calling `router.refresh()`)

**Reply by lynn1286** [reply]
Here I have a question, calling the cookies function results in dynamic rendering, does this mean that my whole project loses static rendering and is the unstable_setRequestLocale method still available?

**Reply by theoludwig** [reply]
Is it possible to support a use case where pages are statically rendered at build time for every locales, still no i18n routing and the locale of the user is determined dynamically with cookies, but as the page has already been built with the locale, it's really fast. Basically same question as @lynn1286

**Reply by amannn** [reply]
Reading from `cookies()` means you opt out of static rendering, that's correct—yes.

> Is it possible to support a use case where pages are statically rendered at build time for every locales, still no i18n routing and the locale of the user is determined dynamically with cookies, but as the page has already been built with the locale, it's really fast

That's very much what the `localePrefix: 'never'` strategy does. If you're looking for this, that setting might be the right option for you.

**Reply by theoludwig** [reply]
Thank you for the fast answer.

> Reading from cookies() means you opt out of static rendering, that's correct—yes.

That's exactly th problem. I wish there could be a way to mix top level layout dynamic (set `<html lang={lang}>` from cookie) while preserving static rendering for the rest of the page (children of the layout), as all the possible "dynamic" pages were already rendered, so the server side for each request, only read the cookie, validate it (probably in middleware), set the `lang` to the html tag, and that's all that is different for each request, then the rest is completely static, it only responds with the corresponding rendered html (generated at build time) for the given lang.

I think, the solution for this would be to share a value (the locale/lang) between `middleware.ts`, and the pages/components, without using `cookies()`, `headers()` functions so it doesn't opt out of static rendering.

**Reply by theoludwig** [reply]
Oh actually, I might be misunderstanding, and it's actually working/it is supported?

I'm checking the examples, and tried `example-app-router`, I've only changed `export const localePrefix = undefined;` to `export const localePrefix = 'never' as const;` in `src/navigation.ts`. With this, does that mean, pages are still static, without i18n routing, but still dynamic lang with cookie?

**Edit:** From my experimentation, it should indeed work, thank you so much for making this library.

**Reply by amannn** [reply]
> does that mean, pages are still static, without i18n routing, but still dynamic lang with cookie?

Yep, exactly! Internally you have separate routes as defined by the `[locale]` segment and the middleware will rewrite incoming requests accordingly to pick the correct (statically) rendered page.

I think I should add this to the `localePrefix: 'never'` docs, it's actually a good reason to pick this strategy.

**Reply by askjervold** [reply]
> How are you storing your locale? In a cookie, like in the example above? You could try to invoke a Server Action that updates your cookie and then calls revalidatePath('/'). Theoretically, that should return updated markup for the page without a reload.

There's a bit of a delay, but this worked! Thanks for the help! 🤩

**Reply by askjervold** [reply]
A bit of change in requirements has introduced a new conundrum in my use-case. When the user has not explicitly set their preferred language (and also possibly if their preferred language is not available), we want to make an API call based on a dynamic route segment to determine the default and available languages and use this to set the locale. Is this possible to do from the `getRequestConfig` call, or would I need to introduce a middleware to perform this check and set a cookie that I can read from in `getRequestConfig`? Also, if I understand correctly, the config is cached – but is it cached per request path, so that a request to a different path would check the appropriate data based on the new dynamic segment and possibly return a new locale?

**Reply by amannn** [reply]
Hmm, yes, sounds like a job for a middleware as cookies can not be set during the RSC render.

Note however that if you set a cookie on the _response_ in the middleware, you will not be able to read it from a _request_ header during the RSC render. A workaround is to attach the value as an additional request header in the middleware that can be read during the RSC render.

You could have a look at the [`localePrefix: 'never'`](https://next-intl-docs.vercel.app/docs/routing#locale-prefix-never) too in case this is a good fit for your requirements. The middleware can be [composed](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares) to dynamically assign the locales.

> the config is cached – but is it cached per request path, so that a request to a different path would check the appropriate data based on the new dynamic segment and possibly return a new locale?

See [`i18n.ts` in the docs](https://next-intl-docs.vercel.app/docs/usage/configuration#i18nts)—the cache rules from React's `cache` apply here (i.e. the cache is request-scoped, requests for the same path will call this code again—at least if you're using dynamic rendering, e.g. caused by reading headers).
